.gitlab-ci.yml

```diff
@@ -0,0 +1,95 @@
+---
+image: docker:latest
+services:
+  - docker:dind
+
+variables:
+  DOCKER_HUB_REGISTRY_IMAGE: codeception/codeception
+  ALPINE_TAG: $CI_BUILD_REF_NAME-alpine
+  ALPINE_IMAGE_TEST: $CI_REGISTRY_IMAGE:$CI_BUILD_REF_NAME-alpine-test
+
+stages:
+  - build
+  - test
+  - deploy
+
+.dind_before_script: &dind_before_script
+  before_script:
+    - docker login -u gitlab-ci-token -p $CI_BUILD_TOKEN $CI_REGISTRY
+
+composer_install:
+  image: php:7.0-alpine
+  stage: build
+  script:
+    - apk update --no-cache
+    - apk add --no-cache git zlib-dev openssl
+    - docker-php-ext-install zip
+    - curl -sS https://getcomposer.org/installer | php -- --filename=composer --install-dir=/usr/local/bin
+    - composer global require --optimize-autoloader "hirak/prestissimo"
+    - composer install --prefer-dist --optimize-autoloader --ignore-platform-reqs
+  artifacts:
+      expire_in: 1 hour
+      when: on_success
+      paths:
+        - vendor/
+
+docker_build:alpine:
+  <<: *dind_before_script
+  stage: build
+  script:
+    - >
+      docker build --rm
+      --tag $ALPINE_IMAGE_TEST
+      --file Dockerfile-alpine .
+    - docker push $ALPINE_IMAGE_TEST
```

Comment:
you should:
- create docker image tag in gitlab.com registry named by branch+commit hash
- pull that in test stage
- tag & push the final image in publish stage

see this excellent example:
https://gitlab.com/markpundsack/docker-example